File: frontend/src/shared/notifications/studyReminders.js
import { fetchPlannerAgenda } from '../api/workspace.api.js';
import {
  showLocalNotification,
  cancelLocalNotifications,
  getLocalNotificationPermission,
} from '../platform/native/LocalNotifications.js';

/**
 * On-device study reminders. Two sources:
 *   1. Planner agenda due-dates  -> remind N hours before each open task/exam.
 *   2. User-set custom times     -> a daily reminder at a chosen HH:MM.
 *
 * Everything is scheduled locally on the device (no server push). Preferences live in
 * localStorage. Reminder ids are derived deterministically so re-reconciling cancels the
 * previous batch and re-schedules cleanly without duplicates.
 */

const PREFS_KEY = 'lms_study_reminder_prefs';
const SCHEDULED_IDS_KEY = 'lms_study_reminder_ids';

// Deterministic id ranges so we only cancel/replace our own scheduled reminders.
const PLANNER_ID_BASE = 700_000_000;
const PLANNER_ID_SPAN = 100_000;
const CUSTOM_ID = 690_000_001;

const DEFAULT_PREFS = {
  plannerEnabled: true,
  plannerLeadHours: 3,
  customEnabled: false,
  customTime: '18:00', // HH:MM, local
};

export function getStudyReminderPrefs() {
  if (typeof window === 'undefined') return { ...DEFAULT_PREFS };
  try {
    const parsed = JSON.parse(window.localStorage.getItem(PREFS_KEY) || '{}');
    return { ...DEFAULT_PREFS, ...parsed };
  } catch {
    return { ...DEFAULT_PREFS };
  }
}

export function saveStudyReminderPrefs(next) {
  if (typeof window === 'undefined') return getStudyReminderPrefs();
  const merged = { ...getStudyReminderPrefs(), ...next };
  window.localStorage.setItem(PREFS_KEY, JSON.stringify(merged));
  return merged;
}

function plannerReminderId(index) {
  return PLANNER_ID_BASE + (index % PLANNER_ID_SPAN);
}

/** Ids we actually scheduled last time, so we only cancel our own batch. */
function readScheduledIds() {
  if (typeof window === 'undefined') return [];
  try {
    const parsed = JSON.parse(window.localStorage.getItem(SCHEDULED_IDS_KEY) || '[]');
    return Array.isArray(parsed) ? parsed.filter((id) => Number.isInteger(id)) : [];
  } catch {
    return [];
  }
}

function writeScheduledIds(ids) {
  if (typeof window === 'undefined') return;
  window.localStorage.setItem(SCHEDULED_IDS_KEY, JSON.stringify(ids));
}

function parseDueDate(item) {
  const raw =
    item?.dueAt || item?.dueDate || item?.due_at || item?.due_date || item?.date || item?.startAt || '';
  if (!raw) return null;
  const date = new Date(raw);
  return Number.isNaN(date.getTime()) ? null : date;
}

function nextDailyOccurrence(timeStr) {
  const [h, m] = String(timeStr || '18:00')
    .split(':')
    .map((v) => Number(v));
  if (!Number.isInteger(h) || !Number.isInteger(m)) return null;
  const now = new Date();
  const next = new Date(now);
  next.setHours(h, m, 0, 0);
  if (next.getTime() <= now.getTime()) next.setDate(next.getDate() + 1);
  return next;
}

function isOpenAgendaItem(item) {
  const status = String(item?.status || '').toLowerCase();
  return status !== 'done' && status !== 'completed' && status !== 'archived';
}

/**
 * Reconcile all study reminders against current prefs + planner agenda.
 * Cancels our prior batch, then re-schedules. Safe to call repeatedly.
 */
export async function reconcileStudyReminders() {
  const permission = await getLocalNotificationPermission();
  if (permission !== 'granted') {
    return { ok: false, scheduled: 0, reason: 'permission' };
  }

  const prefs = getStudyReminderPrefs();

  // Clear only the reminders we scheduled previously (native only; no-op on web).
  await cancelLocalNotifications(readScheduledIds());

  const scheduledIds = [];
  let scheduled = 0;

  // 1. Custom daily reminder.
  if (prefs.customEnabled) {
    const at = nextDailyOccurrence(prefs.customTime);
    if (at) {
      await showLocalNotification({
        id: CUSTOM_ID,
        title: 'Study time',
        body: 'Time for your study session. Open your planner to get started.',
        url: '/planner',
        at,
      });
      scheduledIds.push(CUSTOM_ID);
      scheduled += 1;
    }
  }

  // 2. Planner due-date reminders.
  if (prefs.plannerEnabled) {
    const agenda = await fetchPlannerAgenda().catch(() => []);
    const items = Array.isArray(agenda) ? agenda : agenda?.items || [];
    const leadMs = Math.max(0, Number(prefs.plannerLeadHours) || 0) * 60 * 60 * 1000;
    const now = Date.now();

    const upcoming = items
      .filter(isOpenAgendaItem)
      .map((item) => ({ item, due: parseDueDate(item) }))
      .filter(({ due }) => due && due.getTime() - leadMs > now)
      .sort((a, b) => a.due.getTime() - b.due.getTime())
      .slice(0, 32); // cap scheduled reminders

    let index = 0;
    for (const { item, due } of upcoming) {
      const at = new Date(due.getTime() - leadMs);
      const title = item?.title ? `Upcoming: ${item.title}` : 'Upcoming study task';
      const typeLabel = item?.type ? `${item.type} ` : '';
      const id = plannerReminderId(index);
      await showLocalNotification({
        id,
        title,
        body: `Your ${typeLabel}is due soon. Tap to review it in your planner.`,
        url: '/planner',
        at,
      });
      scheduledIds.push(id);
      index += 1;
      scheduled += 1;
    }
  }

  writeScheduledIds(scheduledIds);
  return { ok: true, scheduled };
}

/** Cancel every study reminder we scheduled (native only). */
export async function clearStudyReminders() {
  await cancelLocalNotifications(readScheduledIds());
  writeScheduledIds([]);
}

File: mobile/lib/services/study_reminders.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

import '../features/planner/planner_repository.dart';
import 'notifications.dart';

/// On-device study reminders — mirrors the web LMS `studyReminders.js` workflow:
///   1. Planner due-dates → remind N lead-hours before each open task.
///   2. An optional daily reminder at a chosen HH:MM.
/// Prefs live in SharedPreferences. Reminder ids are deterministic so each
/// reconcile cancels the previous batch and reschedules cleanly (no duplicates).
class StudyReminderPrefs {
  final bool plannerEnabled;
  final int plannerLeadHours;
  final bool customEnabled;
  final String customTime; // 'HH:MM'

  const StudyReminderPrefs({
    this.plannerEnabled = true,
    this.plannerLeadHours = 3,
    this.customEnabled = false,
    this.customTime = '18:00',
  });

  StudyReminderPrefs copyWith({
    bool? plannerEnabled,
    int? plannerLeadHours,
    bool? customEnabled,
    String? customTime,
  }) =>
      StudyReminderPrefs(
        plannerEnabled: plannerEnabled ?? this.plannerEnabled,
        plannerLeadHours: plannerLeadHours ?? this.plannerLeadHours,
        customEnabled: customEnabled ?? this.customEnabled,
        customTime: customTime ?? this.customTime,
      );

  Map<String, dynamic> toJson() => {
        'plannerEnabled': plannerEnabled,
        'plannerLeadHours': plannerLeadHours,
        'customEnabled': customEnabled,
        'customTime': customTime,
      };

  factory StudyReminderPrefs.fromJson(Map<String, dynamic> m) =>
      StudyReminderPrefs(
        plannerEnabled: m['plannerEnabled'] != false,
        plannerLeadHours: (m['plannerLeadHours'] is num)
            ? (m['plannerLeadHours'] as num).toInt()
            : 3,
        customEnabled: m['customEnabled'] == true,
        customTime: (m['customTime'] is String) ? m['customTime'] : '18:00',
      );
}

class StudyReminders {
  static const _prefsKey = 'lms_study_reminder_prefs';
  static const _idsKey = 'lms_study_reminder_ids';
  static const _plannerIdBase = 700000000;
  static const _plannerIdSpan = 100000;
  static const _customId = 690000001;

  static Future<StudyReminderPrefs> getPrefs() async {
    final sp = await SharedPreferences.getInstance();
    try {
      final raw = sp.getString(_prefsKey);
      if (raw == null) return const StudyReminderPrefs();
      return StudyReminderPrefs.fromJson(jsonDecode(raw) as Map<String, dynamic>);
    } catch (_) {
      return const StudyReminderPrefs();
    }
  }

  static Future<void> savePrefs(StudyReminderPrefs p) async {
    final sp = await SharedPreferences.getInstance();
    await sp.setString(_prefsKey, jsonEncode(p.toJson()));
  }

  static DateTime? _nextDaily(String time) {
    final parts = time.split(':');
    if (parts.length != 2) return null;
    final h = int.tryParse(parts[0]), m = int.tryParse(parts[1]);
    if (h == null || m == null) return null;
    final now = DateTime.now();
    var next = DateTime(now.year, now.month, now.day, h, m);
    if (!next.isAfter(now)) next = next.add(const Duration(days: 1));
    return next;
  }

  /// Reconcile all reminders against current prefs + tasks. Safe to call often.
  static Future<int> reconcile(List<PlannerTask> tasks) async {
    final granted = await Notifications.requestPermission();
    if (!granted) return 0;
    final prefs = await getPrefs();
    final sp = await SharedPreferences.getInstance();

    // Cancel only the batch we scheduled last time.
    for (final s in sp.getStringList(_idsKey) ?? const <String>[]) {
      final id = int.tryParse(s);
      if (id != null) await Notifications.cancel(id);
    }

    final scheduled = <int>[];

    // 1. Daily custom reminder.
    if (prefs.customEnabled) {
      final at = _nextDaily(prefs.customTime);
      if (at != null) {
        await Notifications.schedule(
          id: _customId,
          title: 'Study time',
          body: 'Time for your study session — open your planner to get started.',
          when: at,
        );
        scheduled.add(_customId);
      }
    }

    // 2. Planner due-date reminders (lead-hours before end of the due day).
    if (prefs.plannerEnabled) {
      final lead = Duration(hours: prefs.plannerLeadHours.clamp(0, 72));
      final now = DateTime.now();
      final upcoming = tasks
          .where((t) => !t.done && t.due != null)
          .map((t) {
            final d = t.due!;
            final endOfDay = DateTime(d.year, d.month, d.day, 23, 59);
            return (task: t, at: endOfDay.subtract(lead));
          })
          .where((e) => e.at.isAfter(now))
          .toList()
        ..sort((a, b) => a.at.compareTo(b.at));

      var index = 0;
      for (final e in upcoming.take(32)) {
        final id = _plannerIdBase + (index % _plannerIdSpan);
        await Notifications.schedule(
          id: id,
          title: 'Upcoming: ${e.task.title}',
          body: 'This task is due soon. Tap to review it in your planner.',
          when: e.at,
        );
        scheduled.add(id);
        index++;
      }
    }

    await sp.setStringList(_idsKey, scheduled.map((e) => '$e').toList());
    return scheduled.length;
  }
}
